Pixel pack and unpack state is cached but also asked from the underlying context

kkinnunen-apple opened the pull request:
#### cd21603ce53071ee1f26c8bf86fe4761316d41cc
<pre>
Pixel pack and unpack state is cached but also asked from the underlying context
<a href="https://bugs.webkit.org/show_bug.cgi?id=257314">https://bugs.webkit.org/show_bug.cgi?id=257314</a>
rdar://109821711

Reviewed by Dan Glastonbury.

WebGLRenderingContextBase, WebGL2RenderingContext caches the
PACK_*, UNPACK_* state. When the WebGL client asks for the state,
the context will ask the underlying GraphicsContextGL for the state,
even though it is available as cached.

Return the cached state directly.
Use WebKit style for the functions and types related to this: use
full names and avoid using get prefix for accesors.

This is work towards being able to not send the state to the underlying
context. Remote variant should always readPixels with ignoring the client
packing in order to minimize transfers. This will be implemented in
subsequent patches. This means the underlying pack state and WebGL
pack state are going to be different, and thus asking the pack state
from the underlying context would need redundant mutation of the
state.

* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::initializeNewContext):
(WebCore::WebGL2RenderingContext::getTextureSourceSubRectangle):
(WebCore::WebGL2RenderingContext::pixelStorei):
(WebCore::WebGL2RenderingContext::texImage3D):
(WebCore::WebGL2RenderingContext::texSubImage3D):
(WebCore::WebGL2RenderingContext::getParameter):
(WebCore::WebGL2RenderingContext::resetUnpackParameters): Deleted.
(WebCore::WebGL2RenderingContext::restoreUnpackParameters): Deleted.
(WebCore::WebGL2RenderingContext::getPackPixelStoreParams const): Deleted.
(WebCore::WebGL2RenderingContext::getUnpackPixelStoreParams const): Deleted.
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::ScopedTightUnpackParameters::ScopedTightUnpackParameters):
(WebCore::ScopedTightUnpackParameters::~ScopedTightUnpackParameters):
(WebCore::ScopedTightUnpackParameters::set):
(WebCore::WebGLRenderingContextBase::initializeNewContext):
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::pixelStorei):
(WebCore::WebGLRenderingContextBase::texImageSource):
(WebCore::WebGLRenderingContextBase::texImageArrayBufferViewHelper):
(WebCore::WebGLRenderingContextBase::texImageImpl):
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
(WebCore::WebGLRenderingContextBase::unpackPixelStoreParameters const):
(WebCore::ScopedUnpackParametersResetRestore::ScopedUnpackParametersResetRestore): Deleted.
(WebCore::ScopedUnpackParametersResetRestore::~ScopedUnpackParametersResetRestore): Deleted.
(WebCore::WebGLRenderingContextBase::resetUnpackParameters): Deleted.
(WebCore::WebGLRenderingContextBase::restoreUnpackParameters): Deleted.
(WebCore::WebGLRenderingContextBase::getPackPixelStoreParams const): Deleted.
(WebCore::WebGLRenderingContextBase::getUnpackPixelStoreParams const): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::pixelStorePackParameters const):
(WebCore::WebGLRenderingContextBase::unpackPixelStoreParameters const):
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::computeImageSizeInBytes):
(WebCore::GraphicsContextGL::packImageData):
(WebCore::GraphicsContextGL::extractPixelBuffer):
(WebCore::GraphicsContextGL::extractTextureData):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:

Canonical link: <a href="https://commits.webkit.org/264567@main">https://commits.webkit.org/264567@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f8698de2b21a65ba72d3528d4e566e8c7dc82fe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8010 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8295 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8507 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9662 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8014 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10281 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8206 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8154 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9242 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9782 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6543 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7330 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14908 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7666 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7451 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10811 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7922 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6445 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7236 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1910 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11444 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7663 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -3964,7 +3970,7 @@ void WebGLRenderingContextBase::texImageImpl(TexImageFunctionID functionID, GCGL
     }
 
-    ScopedUnpackParametersResetRestore temporaryResetUnpack(this, true);
+    ScopedTightUnpackParameters temporaryResetUnpack(*this, true);
```
Previously (Line 3668) this is written as `ScopedTightUnpackParameters temporaryResetUnpack(*this)`. Remove the `true` or the default parameter to `ScopedTightUnpackParameters` and be explicit everywhere.

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
Pixel pack and unpack state is cached but also asked from the underlying context
https://bugs.webkit.org/show_bug.cgi?id=257314
rdar://109821711

Reviewed by Dan Glastonbury.

WebGLRenderingContextBase, WebGL2RenderingContext caches the
PACK_*, UNPACK_* state. When the WebGL client asks for the state,
the context will ask the underlying GraphicsContextGL for the state,
even though it is available as cached.

Return the cached state directly.
Use WebKit style for the functions and types related to this: use
full names and avoid using get prefix for accesors.

This is work towards being able to not send the state to the underlying
context. Remote variant should always readPixels with ignoring the client
packing in order to minimize transfers. This will be implemented in
subsequent patches. This means the underlying pack state and WebGL
pack state are going to be different, and thus asking the pack state
from the underlying context would need redundant mutation of the
state.

* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::initializeNewContext):
(WebCore::WebGL2RenderingContext::getTextureSourceSubRectangle):
(WebCore::WebGL2RenderingContext::pixelStorei):
(WebCore::WebGL2RenderingContext::texImage3D):
(WebCore::WebGL2RenderingContext::texSubImage3D):
(WebCore::WebGL2RenderingContext::getParameter):
(WebCore::WebGL2RenderingContext::resetUnpackParameters): Deleted.
(WebCore::WebGL2RenderingContext::restoreUnpackParameters): Deleted.
(WebCore::WebGL2RenderingContext::getPackPixelStoreParams const): Deleted.
(WebCore::WebGL2RenderingContext::getUnpackPixelStoreParams const): Deleted.
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::ScopedTightUnpackParameters::ScopedTightUnpackParameters):
(WebCore::ScopedTightUnpackParameters::~ScopedTightUnpackParameters):
(WebCore::ScopedTightUnpackParameters::set):
(WebCore::WebGLRenderingContextBase::initializeNewContext):
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::pixelStorei):
(WebCore::WebGLRenderingContextBase::texImageSource):
(WebCore::WebGLRenderingContextBase::texImageArrayBufferViewHelper):
(WebCore::WebGLRenderingContextBase::texImageImpl):
(WebCore::WebGLRenderingContextBase::validateTexFuncData):
(WebCore::WebGLRenderingContextBase::unpackPixelStoreParameters const):
(WebCore::ScopedUnpackParametersResetRestore::ScopedUnpackParametersResetRestore): Deleted.
(WebCore::ScopedUnpackParametersResetRestore::~ScopedUnpackParametersResetRestore): Deleted.
(WebCore::WebGLRenderingContextBase::resetUnpackParameters): Deleted.
(WebCore::WebGLRenderingContextBase::restoreUnpackParameters): Deleted.
(WebCore::WebGLRenderingContextBase::getPackPixelStoreParams const): Deleted.
(WebCore::WebGLRenderingContextBase::getUnpackPixelStoreParams const): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::pixelStorePackParameters const):
(WebCore::WebGLRenderingContextBase::unpackPixelStoreParameters const):
* Source/WebCore/platform/graphics/GraphicsContextGL.cpp:
(WebCore::GraphicsContextGL::computeImageSizeInBytes):
(WebCore::GraphicsContextGL::packImageData):
(WebCore::GraphicsContextGL::extractPixelBuffer):
(WebCore::GraphicsContextGL::extractTextureData):
* Source/WebCore/platform/graphics/GraphicsContextGL.h:

Canonical link: https://commits.webkit.org/264567@main
